### ISSUE TITLE
MySQL Appliance

### ISSUE THREAD

**Issue by benbjohnson** [flynn/flynn#2364]
## Overview

This is a proposal for implementing a replicated MySQL appliance using [MariaDB Galera Cluster][mariadb-galera].

Galera provides a multi-master clustering implementation which allows reads and writes to any node in the cluster. It handles network partitions, incremental & full snapshots, and eviction so the state machine for MySQL should be significantly simpler than for Postgres.

I propose the following types in `appliance/mysql`:

* `Process` - represents the low-level runnable process and manages configuration writing.

* `Node` - represents the node within the cluster and manages a simple state machine for start/stop. it can also report health of the node based on its cluster membership status from Galera.

[maria-galera]: http://galeracluster.com/documentation-webpages/installmariadb.html

### Bootstrapping

Initializing the cluster will require the `mysql` app to start up with an initial node with the `--wsrep-new-cluster` flag. Subsequent nodes will use the `wsrep_cluster_address` written to the MySQL config file. Once started, they will check the `wsrep_cluster_size` parameter until it shows the new node connected.

### Load Balancing

Because all nodes accept reads and writes, we'll need something to load balance the connections to the multiple MySQL servers (e.g. HAProxy). Should this be included as part of the `mysql` appliance or separate?

### Recovery

Although Galera will recover individual node failures, cluster-wide restarts require manual intervention. Nodes have to be started individually in a recovery mode to determine the most advanced node. The most advanced node must be the first node to be started during the restart.

### Miscellaneous

The Galera documentation provides a small section on ["Container Deployments"](http://galeracluster.com/documentation-webpages/containers.html). There's nothing too surprising here. Just a few extra options for setting the node address and name which are usually derived from the host's network interface.

See also #1540

**Comment by philiplb** [comment]
Nice, but maybe this should be called "MariaDB Appliance" as there are actually some incompatibilities: https://mariadb.com/kb/en/mariadb/mariadb-vs-mysql-compatibility/

**Comment by josephglanville** [comment]
Beyond that I am not sure Galera is the right choice necessarily. Atleast not without more consideration.

Namely Galera when used in full multi-master with writes being directed to all nodes can act very differently to standard MySQL from the application point of view. Specifically transactions that would normally procede fine under normal isolations levels (READ COMMITED) usually can deadlock and most unmodified applications aren't prepared to handle the inevitable ROLLBACK that will be issued on the connection.

Secondly there is the matter of bootstrapping and maintaining consistency even when there are full cluster failures. By using Galera the real "state" of the cluster is stored in the Galera replicated state machine rather than in discoverd which inherently makes it harder for Flynn components to reason about it. See here the cluster fixer and the new monitor I'm working on.

I'm not necessarily saying we shouldn't use Galera, just that we need to think carefully about how to handle these caveats.

The other problem with MySQL though is the seme-sync replication is somewhat of a black sheep, it's not used that commonly and currently doesn't yield very good performance but it's AFTER_SYNC mode would provide the same guarantees as the currently postgres sync replication. Ideally if this worked nicely we could re-use the existing Postgres infrastructure and end up with another highly resilient and stable database.

I also have some reservations from having used Galera in production around it's automatic node eviction and other features that could easily play havoc with clusters if we aren't careful to either disable or tame these features. Same goes for current SST transfer, we will either be stuck with rsync/xtra_backup or need to actually put the work in to roll SST using ZFS snapshots.

**Comment by titanous** [comment]
How much latency difference between `AFTER_SYNC` and Galera? I assume it's not much different? I like the idea of re-using the same state machine.

**Comment by josephglanville** [comment]
Looks to be about the same in versions post 5.6 (I assume we would be using the most recent MariaDB release). Specifically multiple people have cited you can do about 2.5k commits/s using `AFTER_SYNC`. This [blog post](http://yoshinorimatsunobu.blogspot.com.au/2014/04/semi-synchronous-replication-at-facebook.html) goes into some usage of the feature at facebook, though they did end up extending their binlog reciever to support the semi-sync protocol instead of running another full slave (mostly to save CPU more than increase performance of replication).

**Comment by josephglanville** [comment]
There is a few things we need to research still to work out if we can build a Manatee style HA system for MySQL.

1. We need the current transaction log position. This is used to determine if it's safe to do a takeover and declare a new generation.
2. We need to ensure that we can configure MySQL in such a way that if the number of connected semi-sync slaves drops to 0 that the database becomes unwritable rather than falling back to async replication.

Now I am fairly sure the first is pretty achievable. I am however uncertain about the second one.
Unfortunately we need both of these in order to implement the state machine safely.
It could be that we could create a patch for the latter, MariaDB community development is very active and it's likely they would accept a patch that would do this via a setting.

**Comment by josephglanville** [comment]
A cursory look at MariaDB documentation indicates that [SHOW MASTER STATUS](https://mariadb.com/kb/en/mariadb/show-master-status/) and [SHOW SLAVE STATUS](https://mariadb.com/kb/en/mariadb/show-slave-status/) should give us the necessary information regarding the log file position. We will probably also need to incorporate the log file name. This is ok though as long as we are able to sort them.

As for pausing when semi-sync slaves reaches 0 this does exist and it has a configurable timeout before switching to async. We need to work out what a value of 0 does (if this makes it wait forever, i.e good) or if we just set it to the max value which is `18446744073709551615`... which is effectively forever. :laughing: Either way our end goal is achieved, which is to pause transactions on the master if the sync is not acking them. Further docs [here](https://mariadb.com/kb/en/mariadb/semisynchronous-replication/).

So things look good.

As for the actual work required.. We could make the state machine generic by turning `Peer` into an interface and ensuring we have hooks that support stuff like `postgres init`/`mysql_install_db` etc. It would also need to be made responsible for exposing the "WAL/xlog" concept of the underlying database as a monotonically increasing integer for use in the state machine. Apart from the `Peer` interface there would also need to be the actual `Process` style wrapper similar to what @benbjohnson put together for Redis and the existing Postgres one. This would make the state machine applicable then for any datastore that has linearisable history that can be instrumented and synchronous replication (as it doesn't matter if the async is using synchronous replication). It could be extended later to support systems that don't have synchronous replication just it wouldn't guarantee safety when operating in this mode (see Redis, though users could opt into making it safe-ish by using `WAIT` on the client side).

**Comment by benbjohnson** [comment]
@josephglanville @titanous I have a master/slave in semisync. One thing I found out is that setting `rpl_semi_sync_master_timeout` to the max value of `18446744073709551615` causes the server to not start at all. I dropped it down to a year (`31536000000`) and it starts up right away. I'm not sure what the threshold is where it starts working but I couldn't find anything in the logs to show why it wouldn't start up.

Unfortunately, I'm getting some strange behavior where it's replicating the data but the `mysql` CLI is hanging indefinitely. The data gets replicated to the slave and I can query it there. When I hit `CTRL-C` to cancel the query it persist the query to the master as well. I have no idea what it's waiting for. Here's the output:

```
MariaDB [foo]> INSERT INTO T (name) VALUES ('N');
^CCtrl-C -- query killed. Continuing normally.
Query OK, 1 row affected (57.88 sec)
```

I tried looking at the general log but there's nothing useful. I'm trying to get a trace running in debug mode but no luck yet. I also tried `strace` out of desperation but it's not showing anything while I'm querying. Do you guys have any ideas?

**Comment by josephglanville** [comment]
I'm having trouble parsing `it persist the query to the master as well`, do you mean you attempted an insert on the slave and it hung and did something to the master?

I will prob setup MariaDB and start playing around with this too. The other stuff I am working on feels really close to being working.

**Comment by benbjohnson** [comment]
@josephglanville I ran `INSERT INTO T (name) VALUES ('N');` in a `mysql` CLI client connected to master. The insert gets propagated to the slave and I can query it on the slave. It gets committed to the master as well and I can query it on the master in a separate `mysql` client. However, the CLI hangs. When I hit `CTRL-C`, it says, `query killed` but the insert is still committed. It's weird for sure.

I'll continue on the appliance code itself for now. I'm assuming this is just an odd issue. Let me know if you have any problems getting it up and running.

**Comment by josephglanville** [comment]
Ok, will keep my eyes out. Thanks for clarification. :)

**Comment by titanous** [comment]
Implemented in #2433.
